### ISSUE TITLE
Diagram - Data Binding API Enhancements

### ISSUE THREAD

**Issue by dxbykov** [DevExpress/DevExtreme#14354]
# The Problem

Currently, the Diagram does not support partial updates based on data source changes.

1. When diagram data is changed outside of the component (for example, when you modify shape data in a custom edit form), these changes are not reflected in the diagram’s model. The only way to update Diagram shapes and connectors is to reload all data. This is a time-consuming process that may lead to UX issues.
1. The same limitation affects scenarios with predefined configuration options for a new shape added from the toolbox. Partial upload is also necessary when a shape is copied - to apply the original shape data to the copied shape.
1. The Diagram cannot pass additional data source fields that are not mapped to its shape template options. As a result, it is not possible to copy shapes with custom data fields or use the component’s Undo/Redo commands to undo/redo these custom data modifications.

# The Proposed Solution

## Partial Updates

We introduced a special `reloadContent(itemKey, applyLayout)` method to update Diagram data based on data source changes on demand. This method reloads the modified diagram elements from the data source and recreates the diagram layout.

*Note that this is a temporary API which can be replaced with the built-in reload functionality. It will not require calling additional methods in the official 20.2 release.*

The `itemKey` parameter specifies the key of the item that should be reloaded. If this parameter is not specified, all diagram items will be updated. The `applyLayout` parameter defines whether or not the corresponding layout should be applied to the Diagram after reload.

For example, this method can be called in Angular to reload a specific Diagram shape based on its key in the ArrayStore’s Updated event handler:      

```
onUpdated: function(key, values) {
     this.diagram.instance.reloadContent(key, true);
},
```

## Enhanced Change Tracking

We implemented the `nodes.customDataExpr` and `edges.customDataExpr` options. These options help link custom data from a data source to diagram elements, so that you can track changes in the component’s history. 

This API includes the following capabilities:

1. Display extra data (data source fields) in shape templates and reflect their changes in the diagram’s model and history. This way, changes in this extra data can be reverted and applied with the control’s Undo/Redo option.
1. Persist custom fields’ values when a user copies and pastes a diagram element. You can initialize the pasted element’s data item object with extra fields' values of the original shape. For example:

```
<dxo-nodes [dataSource]="dataSource" keyExpr="ID" " [customDataExpr]="itemCustomDataExpr">
```

```
   itemCustomDataExpr(obj, value) {
        if(value === undefined) {
            return {
              "FullName": obj.FullName,
              "Prefix": obj.Prefix,
            };
        } else {
            obj.FullName = value.FullName;
            obj.Prefix = value.Prefix;
        }
    }
```

# Try It

- New Toolbox Configuration Options ([jQuery](https://codepen.io/roman-rodin/pen/QWNdzEY?editable=true%3Dhttp%3A%2F%2Fjsserver%3A81%2FDemos-20.2%2FWidgetsGallery%2FDemo%2FDiagram%2FUICustomization%2FjQuery%2FLight%2F))
- Partial updates and data binding enhancements ([jQuery](https://codepen.io/roman-rodin/pen/yLOgGgm?editable=true%3Dhttp%3A%2F%2Fjsserver%3A81%2FDemos-20.2%2FWidgetsGallery%2FDemo%2FDiagram%2FCustomShapesWithTemplatesWithEditing%2FjQuery%2FLight%2F))

# We Need Your Feedback

## Take a Quick Poll

[Do you find the new Diagram data binding API useful?](https://docs.google.com/forms/d/e/1FAIpQLSeVl9-te5bKEPIdC8qEq4f3guDuEuaXU2CD2_yz0mEsW2W6Fg/viewform)

# Get Notified of Updates

Subscribe to this thread or to our [Facebook](https://www.facebook.com/DevExpress.DevExtreme/) and [Twitter](https://twitter.com/devextreme) accounts for updates on this topic.

**Comment by COBRASoft5** [comment]
I have a question... Will the diagram 'reset' its visual position after adding external elements?

**Comment by Romzay** [comment]
Hello,
Now you do not need to reload the diagram's data source after adding external elements to it. So, diagram items' position will not be reset.

**Comment by TheReal-BiffBoff** [comment]
Just wanted to try this feature in a Sanbox using the diagram for vue. I was not able to find the reload method on the diagram instance. Is this not implemented for vue yet? 

The partial updating feature would be important for our project and I'd really like to test it :)

**Comment by LanetteDevEx** [comment]
Hello,

Note that the "reloadContent" method was a temporary API, which we replaced with the built-in reload functionality in v20.2. In this version, it is not necessary to explicitly call the "reloadContent" method to update Diagram elements based on data source changes. Two-way data binding is implemented internally, so this reloading process is executed automatically without any Diagram methods.

The Diagram UI component reloads modified diagram data each time the diagram API or data source’s [push API](https://js.devexpress.com/Documentation/Guide/Data_Binding/Data_Layer/#Data_Modification/Integration_with_Push_Services) is used to change the data source.

**Comment by dxbykov** [comment]
Thanks to everybody who gave us feedback on this feature. The feature is available in v20.2. I'm closing this thread. In the case of bugs or questions, feel free to create a new ticket in our [Support Center](https://www.devexpress.com/Support/Center/).
